### ISSUE TITLE
memory leak in `ModelSyncedEventEmitter`

### ISSUE THREAD

**Issue by lawmicha** [aws-amplify/amplify-swift#3510]
Also I found a small leak in `ModelSyncedEventEmitter`, you can find a patch for this here: 
[memoryLeak.patch](https://github.com/aws-amplify/amplify-swift/files/14196335/memoryLeak.patch)

This memory leak might be better tracked by another issue but as the code to easily reproduce it is here, I let you organise this as you want.

_Originally posted by @gbitaudeau in https://github.com/aws-amplify/amplify-swift/issues/3259#issuecomment-1932387569_

**Comment by lawmicha** [comment]
```swift
diff --git a/AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/InitialSync/ModelSyncedEventEmitter.swift b/AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/InitialSync/ModelSyncedEventEmitter.swift
index 9a44dbbc4..9bd806267 100644
--- a/AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/InitialSync/ModelSyncedEventEmitter.swift
+++ b/AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/InitialSync/ModelSyncedEventEmitter.swift
@@ -65,7 +65,7 @@ final class ModelSyncedEventEmitter {
         self.syncOrchestratorSink = initialSyncOrchestrator?
             .publisher
             .receive(on: queue)
-            .filter(filterSyncOperationEvent(_:))
+            .filter { [weak self] in self?.filterSyncOperationEvent($0) ?? false }
             .sink(receiveCompletion: { _ in },
                   receiveValue: { [weak self] value in
                     self?.onReceiveSyncOperationEvent(value: value)
@@ -74,7 +74,7 @@ final class ModelSyncedEventEmitter {
         self.reconciliationQueueSink = reconciliationQueue?
             .publisher
             .receive(on: queue)
-            .filter(filterReconciliationQueueEvent(_:))
+            .filter { [weak self] in self?.filterReconciliationQueueEvent($0) ?? false }
             .sink(receiveCompletion: { _ in },
                   receiveValue: { [weak self] value in
                     self?.onReceiveReconciliationEvent(value: value)

```

**Comment by 5d** [comment]
Confirmed the memory leak does exist, and the proposed patch addresses and resolves the issue. Will created a PR for review and run tests.

**Comment by 5d** [comment]
The patch has been released in version [2.33.4](https://github.com/aws-amplify/amplify-swift/releases/tag/2.33.4). Closing this ticket.
